### DISCUSSION TITLE
How do I access variables of QUploader?

### DISCUSSION THREAD

**Discussion by ElHex** [quasarframework/quasar#14961] [General - Components / Directives / etc]
Hey!, I'm using vue composition API, Ive tried everything I can think of trying to get the QUploader component instance variables inside my component script setup:

This is my template file:
```
<template>
  <q-dialog ref="dialogRef">
    <q-card class="q-dialog-plugin" style="width: 700px; max-width: 80vw;">
      <q-form class="q-gutter-md exportForm">
        <div>
          <div class="row">
            <div class="col-4">
              <div>
                <div class="fileUploaderContainer">
                  <q-uploader ref="fileUploader" class="fileUploader"  url="http://localhost:4444/upload" :style="{'max-width': '250px'}"/>
                  <div class="fileUploadInitializer" style="width: 200px; height: 100px; background: #f00;" @click="startUploader($event)" >

                  </div>
                </div>
              </div>
            </div>
            <div class="col-8">col-8</div>
          </div>
        </div>
      </q-form>
    </q-card>
  </q-dialog>
</template>

<script lang="ts" setup>
import { QUploader, useDialogPluginComponent } from 'quasar';
import { nextTick, onMounted, ref } from 'vue';

const fileUploader = ref(null)
function startUploader(event: MouseEvent){
  console.log('test');
  //fileUploader.value.pickFiles(event)
}

defineEmits([...useDialogPluginComponent.emits]);

onMounted(() => {
  console.log('ty');


    console.log(fileUploader.value);


})
```

I wish to access the "files" array inside the ref "fileUploader" but in the "onMounted" event its null, I've tried ref, defineExpose(fileUploader) and it does not want to work

maybe this is a bug?

I want to hide this component if the files array is empty, thats my goal, but haven't been able to access it

I'm using quasar 2.9.2

**Comment by yusufkandemir** [comment]
For custom dialog plugin components, you should use `@show` on `q-dialog` instead of `onMounted` if you want to access children of `q-dialog`. Either way, I am pretty sure the files will be empty because the user has no time to add files when the dialog is literally just opened. So, you might want to revisit the logic.

Btw, you are missing `@hide="onDialogHide"` and the `useDialogPluginComponent()` call. So, your custom dialog plugin will not work properly.

**Reply by ElHex** [reply]
yeah I wish to have the empty array on load, but I haven't found a wait to access it

**Reply by yusufkandemir** [reply]
You said
> I want to hide this component if the files array is empty, thats my goal

and it will be always empty, so you would have to close it every time it opens, which doesn't make sense to me.

I already answered how to access it, but I guess I need to provide you with more code:
```html
<template>
  <q-dialog ref="dialogRef" @hide="onDialogHide" @show="onShow">
    ...
  </q-dialog>
</template>

<script lang="ts" setup>
import { QUploader, useDialogPluginComponent } from 'quasar';
import { ref, Ref } from 'vue';

// ...
const fileUploader = ref() as Ref<QUploader>;

function onShow () {
  console.log(fileUploader.value);
}
</script>
```

**Reply by ElHex** [reply]
Sorry, maybe I'm not explaining myself properly? I understand what you explained, basically I'm trying to have something like this:

`<q-uploader ref="fileUploader" v-show="fileUploader?.value?.files?.length>0" class="fileUploader"  url="http://localhost:4444/upload" :style="{'max-width': '250px'}"/>`

I'm trying to avoid having an extra variable that tracks the files inside that child component (and the only way I've been able to access it its with the watcheffect function)

```
 watchEffect(() => {
    console.log(fileUploader.value.files);
  })
```

notice the v-show argument, I get it about the dialog ref and the @hide and @show, but, I'm trying to find a way to access 

`fileUploader?.value?.files?.length` even if the array its empty on the setup, so this way the fileUploader always has the instance of the child component, in this case its a QUploader component

Thank you very much for your help

**Reply by yusufkandemir** [reply]
I don't have any clue whatsoever about the stuff you are talking about. You've pretty much sent the same answer to both of us, so I guess there is a problem. I will just take a wild guess:
```js
const files = computed(() => fileUploader.value?.files || [])
```

**Reply by ElHex** [reply]
@yusufkandemir just tried this, I didnt think of using it this way, but the array still does not change, even after adding a file to the uploader, the array is still empty.

**Reply by yusufkandemir** [reply]
There seems to be a problem with computedProps(`files`, `queuedFiles`, etc.) of QUploader: #14967

**Reply by yusufkandemir** [reply]
Temporarily, the only way to get it is to maintain a list yourself by listening to `@added` and `@removed` as shown by @suleimanas

**Reply by ElHex** [reply]
thank you @yusufkandemir, I tried everything but didnt test if the array was being updated on quasars end or not, so maybe most of the things I did might have worked haha

hopefully its fixed shortly and I'm really gratefull for your help @suleimanas / @yusufkandemir

**Comment by suleimanas** [comment]
you can use **@added** event

```
<q-uploader ... @added="onAdd" ... />

function onAdd(files){
  // your selected files array
}
```

**Reply by ElHex** [reply]
yeah that works, but I need to access it on load not on a function

**Reply by suleimanas** [reply]
onload you should have an empty array then add\remove the files from the array

```
<q-uploader ... @added="onAdd" @removed="onRemove" ... />

const files = ref([]);

function onAdd(files){
  // 
}

function onRemove(files){
  // 
}
```

And if you are uploading the files yourself **QFile** would be better alternative you can use **v-model**

**Reply by ElHex** [reply]
yeah thats on theory, but I don't know if its quasar's problem or my code that it simply wont work onload (setup), the variable its undefined even using
```
  onMounted(() => {
    if (fileUploader.value) {
    console.log(fileUploader.value);
  } else {
    // not mounted yet, or the element was unmounted (e.g. by v-if)
  }
  })
```

the variable isnt simply defined, only way I've found out to be available its using:
```
  watchEffect(() => {
    console.log(fileUploader.value.files);
  })
```

which kinda works for me becasue I can simply just update a ref from my side everytime the uploader changes, but I guess this isnt a proper way of doing this


my goal its to make something like this to work:
`<q-uploader ref="fileUploader" v-show="fileUploader?.value?.files?.length>0" class="fileUploader"  url="http://localhost:4444/upload" :style="{'max-width': '250px'}"/>`

notice the v-show conditional I'm trying to add user the ref

**Reply by suleimanas** [reply]
Actually it is a proper way of doing it, there is no one right answer.
watchEffect is one of the proper ways, so you can do something like:

```
const isset = ref(false);

watchEffect(() => {
    isset.value = fileUploader?.value?.files?.length>0;
});
```

and check out @yusufkandemir 's answer you can do the same with computed
NOTE: you're making the both of us confused :D

**Reply by ElHex** [reply]
@suleimanas on thing I just noticed is that we all might be right? this seems like a bug on quasar's side, using vue dev tools I'm trying this:
![image](https://user-images.githubusercontent.com/52222397/203606267-37d7b581-1166-4d58-acfa-487d3923aef3.png)
(id its 129)

then I add a file:
![image](https://user-images.githubusercontent.com/52222397/203606405-7d00b145-ae62-4859-a493-af45842cde03.png)

![image](https://user-images.githubusercontent.com/52222397/203606654-3f67c750-7534-43ca-935e-3496985950d4.png)

It is not updating the files array :/

I used the methods both of you gave me and I think its because of this

**Reply by brunobydesign** [reply]
Hi John,

I have no idea whether it is a good or bad practice, but in order to gain full control of the files list in QUploader I had to use the trick below 

- first use a slot for the list
- then make the slot scope available to the code with a variable

HTH

```
<script setup>
  ...
const uploaderList = ref([]);
 ...
</script>

<template>
    <q-uploader>
    ...
        <template v-slot:list="scope">
            {{ uploaderList = scope }} // makes the uploaderList available to the code
            <q-list v-if="scope.files.length > 0">
              <q-item
                v-for="file in scope.files"
                :key="file.__key"
              >
                ...
              </q-item>
            </q-list>
          </template>
    </q-uploader>
</template>
```
